test/unit/params_ref.cpp

```diff
@@ -276,6 +276,141 @@ struct params_ref_test
+                ++it;
+                BOOST_TEST(it == params.end());
+            }
+
+            // one element with empty value
+            {
+                auto r = boost::urls::parse_relative_ref("?=");
+                if(!BOOST_TEST(r.has_value()))
+                    return;
+                url u = *r;
+                params_ref params = u.params();
+                BOOST_TEST_EQ(params.size(), 1);
+                auto it = params.begin();
+                BOOST_TEST(it != params.end());
+                BOOST_TEST((*it).key.empty());
+                BOOST_TEST((*it).has_value == true);
+                BOOST_TEST((*it).value.empty());
+                ++it;
+                BOOST_TEST(it == params.end());
+            }
+
+            // one element with empty value with fragment
+            {
+                // issue #864
+                auto r = boost::urls::parse_relative_ref("?=#");
+                if(!BOOST_TEST(r.has_value()))
+                    return;
+                url u = *r;
+                params_ref params = u.params();
+                BOOST_TEST_EQ(params.size(), 1);
+                auto it = params.begin();
+                BOOST_TEST(it != params.end());
+                BOOST_TEST((*it).key.empty());
+                BOOST_TEST((*it).has_value == true);
+                BOOST_TEST((*it).value.empty());
+                ++it;
+                BOOST_TEST(it == params.end());
+            }
+
+            // one param
+            {
+                auto r = boost::urls::parse_relative_ref("?key=value");
+                if(!BOOST_TEST(r.has_value()))
+                    return;
+                url u = *r;
+                params_ref params = u.params();
+                BOOST_TEST_EQ(params.size(), 1);
+                auto it = params.begin();
+                BOOST_TEST(it != params.end());
+                BOOST_TEST((*it).key == "key");
+                BOOST_TEST((*it).has_value == true);
+                BOOST_TEST((*it).value == "value");
+                ++it;
+                BOOST_TEST(it == params.end());
+            }
+
+            // two params
+            {
+                auto r = boost::urls::parse_relative_ref("?key1=value1&key2=value2");
+                if(!BOOST_TEST(r.has_value()))
+                return;
+                url u = *r;
+                params_ref params = u.params();
+                BOOST_TEST_EQ(params.size(), 2);
+                auto it = params.begin();
+                BOOST_TEST(it != params.end());
+                BOOST_TEST((*it).key == "key1");
+                BOOST_TEST((*it).has_value == true);
+                BOOST_TEST((*it).value == "value1");
+                ++it;
+                BOOST_TEST(it != params.end());
+                BOOST_TEST((*it).key == "key2");
+                BOOST_TEST((*it).has_value == true);
+                BOOST_TEST((*it).value == "value2");
+                ++it;
+                BOOST_TEST(it == params.end());
```

Comment:
It would be cool if lines 398 through 411 could be reduced to this one line:
```
check_eq( u.params(), {
  { "key1", true, "value1" },
  { "key2", true, "value2" } } );
```

Maybe in a separate PR? This would simplify a ton of test code...

Reply:
I think I defined something like std::ranges::equal in one of these tests. Let me check. We could move it to some test_utils.hpp or something.